src/Symfony/Component/Routing/Generator/CompiledUrlGenerator.php

```diff
@@ -53,6 +52,16 @@ public function generate($name, $parameters = [], $referenceType = self::ABSOLUT
 
         list($variables, $defaults, $requirements, $tokens, $hostTokens, $requiredSchemes) = $this->compiledRoutes[$name];
 
+        if (isset($defaults['_canonical_route']) && isset($defaults['_locale'])) {
+            if (\in_array('_locale', $variables, true)) {
```

Comment:
if (!\in_array()) unset() elseif ... 
this would reduce indentation by one level :)
(same below)